### PR TITLE
[16.0][MIG] product_multi_company: Migration to version 16.0

### PR DESCRIPTION
**Pull request by pilarvargas-tecnativa**
Superseed #459 because the PR is very old and the creator stopped taking care of it, this PR recovers the changes made in its migration but following all the steps of a conventional migration and finalising the migration.

cc @Tecnativa TT48972

@pedrobaeza @carolinafernandez-tecnativa please review

### CODE REVIEW

File: product_multi_company/readme/USAGE.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+On the product form view, go to the "Information" tab, and put the companies
```
Comment (carolinafernandez-tecnativa) [inline product_multi_company/readme/USAGE.rst]: tab "Information" renamed for "General Information"

File: product_multi_company/models/product_category.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    route_ids = fields.Many2many(
+        "stock.route",
+        "stock_location_route_categ",
+        "categ_id",
+        "route_id",
+        "Routes",
+        domain=lambda self: [
+            "|",
+            ("company_id", "=", False),
+            ("company_id", "in", self.env.companies.ids),
```
Comment (carolinafernandez-tecnativa) [inline product_multi_company/models/product_category.py]: better use [('company_id', 'in', self.env.companies.ids + [False])]

File: product_multi_company/models/product_category.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+
+class ProductCategory(models.Model):
+    _inherit = "product.category"
+
+    total_route_ids = fields.Many2many(
+        domain=lambda self: [
+            "|",
+            ("company_id", "=", False),
+            ("company_id", "in", self.env.companies.ids),
```
Comment (carolinafernandez-tecnativa) [inline product_multi_company/models/product_category.py]: better use [('company_id', 'in', self.env.companies.ids + [False])]

File: product_multi_company/hooks.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+def post_init_hook(cr, registry):
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    # Change access rule
+    rule = env.ref("product.product_comp_rule")
+    rule.write(
+        {
+            "domain_force": (
+                "['|', ('company_ids', 'in', company_ids),"
+                "('company_ids', '=', False)]"
+            ),
```
Comment (carolinafernandez-tecnativa) [inline product_multi_company/hooks.py:25]: same here

**Comment by carolinafernandez-tecnativa** [comment]
Functional review LGTM :)

Please rebase and check for CI.
Do not forget to squash administrative commits. Thanks!

**Comment by amkarthik** [comment]
@carolinafernandez-tecnativa 

We are getting this issue in this PR:
https://github.com/OCA/multi-company/issues/252

Please check the attached video below and that is recorded in PR runboat.

https://github.com/OCA/multi-company/assets/12694917/024b9513-779c-4266-9e83-8a5e8d6591a8

Thanks

**Comment by pilarvargas-tecnativa** [comment]
> We are getting this issue in this PR: #252

It may have been something transient in the runbot, since duplicating the user as ‘User B’ and repeating the same operation does not cause this problem.

File: product_multi_company/models/product.py [hunk 5]
```diff
@@ -1,19 +0,0 @@
-# Copyright 2023 Karthik <karthik@sodexis.com>
-# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
-
-from odoo import fields, models
-
-
-class ProductProduct(models.Model):
-    _inherit = "product.product"
-
-    company_ids = fields.Many2many(
```
Comment (pedrobaeza) [inline product_multi_company/models/product.py]: Why removing this?

Comment (pilarvargas-tecnativa) [reply product_multi_company/models/product.py]: It was not in v15, it is not necessary

Comment (pilarvargas-tecnativa) [reply product_multi_company/models/product.py]: The fact is that by extending the abstract model in product.template, product.product acquires that field.

**Comment by amkarthik** [comment]
> > We are getting this issue in this PR: #252
> 
> It may have been something transient in the runbot, since duplicating the user as ‘User B’ and repeating the same operation does not cause this problem.

Hi I tested again with the User B. (changed the password of userb to userb).
Same issue happens to me.
Please check.


https://github.com/OCA/multi-company/assets/12694917/7c930b43-2e2a-4a04-8f16-6db26143eda7

File: product_multi_company/models/product.py (hunk 5 above)
Comment (amkarthik) [reply product_multi_company/models/product.py]: @pilarvargas-tecnativa 
From what I know, the above concept is the reason for the user A and user B issue.
As product.product doesn't hold the company_ids directly in it, if someone tries to access the product.product then it tries to compute it from the product.template.
This causing the above issue as the user doesn't have access to all the companies that is set on the product.template.

**Comment by pilarvargas-tecnativa** [comment]
> Hi I tested again with the User B. (changed the password of userb to userb). Same issue happens to me. Please check.

I have retested the test locally, checking in shell the data contained in the bd if it contains the company_ids field. 
image
image

I have replicated the error but it only happens in the case of new companies created, it does not happen with the demo companies already defined in odoo. I am still investigating the error.

**Comment by amkarthik** [comment]
> > Hi I tested again with the User B. (changed the password of userb to userb). Same issue happens to me. Please check.
> 
> I have retested the test locally, checking in shell the data contained in the bd if it contains the company_ids field. image image
> 
> I have replicated the error but it only happens in the case of new companies created, it does not happen with the demo companies already defined in odoo. I am still investigating the error.

Hi,
Can you please confirm the user that you are using in the shell?

**Comment by pilarvargas-tecnativa** [comment]
@amkarthik Actually, the access rule to the ‘product.product’ model was not being applied and it is necessary to do it as it was done in v13 and v14. This same error already occurs in v15 where this part of the code has been removed. This error should no longer appear and I have added the corresponding unit tests so that it does not happen again. Thanks

File: product_multi_company/models/product_category.py (hunk 2 above)
Comment (pilarvargas-tecnativa) [reply product_multi_company/models/product_category.py]: thanks

**Comment by carolinafernandez-tecnativa** [comment]
Please check CI, still having problems with multi-company rules, thanks!

**Comment by amkarthik** [comment]
> @amkarthik Actually, the access rule to the ‘product.product’ model was not being applied and it is necessary to do it as it was done in v13 and v14. This same error already occurs in v15 where this part of the code has been removed. This error should no longer appear and I have added the corresponding unit tests so that it does not happen again. Thanks

Thanks

File: product_multi_company/models/product_product.py
Comment (amkarthik) [inline product_multi_company/models/product_product.py:1]: @pilarvargas-tecnativa 
Why we are going back to using create and write method as in v13.
Why don't we consider creating related field instead?
    company_ids = fields.Many2many(
        comodel_name="res.company",
        column1="product_id",
        column2="company_id",
        relation="product_product_company_rel",
        related="product_tmpl_id.company_ids",
        compute_sudo=True,
        readonly=False,
        store=True,
    )

This is in the PR #459

Comment (pilarvargas-tecnativa) [reply product_multi_company/models/product_product.py:1]: Having debugged all the code after all the conflicts with tests of other modules that uncovered the errors that occurred in this one, it is certain that this option is the most appropriate. The tests to check that the variants have their access permissions correctly set are also maintained. Thanks

**Comment by carolinafernandez-tecnativa** [comment]
With these changes, when creating new companies and accessing with created users to product variant the error seems to disappear.
I also have checked accessing in product template and now it seems to work fine, Thanks !

**Review by carolinafernandez-tecnativa — APPROVED** [review]

File: product_multi_company/__manifest__.py [hunk 6]
```diff
@@ -7,10 +7,12 @@
     "author": "Tecnativa," "Odoo Community Association (OCA)",
     "website": "https://github.com/OCA/multi-company",
     "category": "Product Management",
-    "version": "15.0.1.0.0",
+    "version": "16.0.1.0.0",
     "license": "AGPL-3",
     "depends": ["base_multi_company", "product", "stock"],
-    "data": ["views/product_template_view.xml"],
+    "data": [
+        "views/product_template_view.xml",
```
Comment (pedrobaeza) [inline product_multi_company/__manifest__.py]: This change is not needed. Please undo in the corresponding commit.

File: product_multi_company/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+{
+    "name": "Product multi-company",
+    "summary": "Select individually the product template visibility on each " "company",
+    "author": "Tecnativa," "Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/multi-company",
+    "category": "Product Management",
+    "version": "16.0.1.0.0",
+    "license": "AGPL-3",
+    "depends": ["base_multi_company", "product", "stock"],
```
Comment (pedrobaeza) [inline product_multi_company/__manifest__.py]: The dependency on stock must be removed.

File: product_multi_company/models/product_category.py [hunk 8]
```diff
@@ -8,22 +8,14 @@ class ProductCategory(models.Model):
...
-            "|",
-            ("company_id", "=", False),
-            ("company_id", "in", self.env.companies.ids),
-        ]
+        domain=lambda self: [("company_id", "in", [False] + self.env.companies.ids)]
     )
 
     route_ids = fields.Many2many(
-        "stock.location.route",
+        "stock.route",
```
Comment (pedrobaeza) [inline product_multi_company/models/product_category.py]: Remove these overrides. That should go if ever done on a module `product_multi_company_stock`.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: product_multi_company/models/product_category.py (hunk 8 above)
Comment (pilarvargas-tecnativa) [reply product_multi_company/models/product_category.py]: All done! thanks

**Review by pedrobaeza — APPROVED** [review]
/ocabot merge nobump

**Review by amkarthik — APPROVED** [review]

**Comment by cyrilmanuel** [comment]
Hello @pedrobaeza, could you merge this PR ?

**Pull request merged**